shapeshift/web
# feat: revert of the revert of feat: dashboard qr code button (#4378)

## Description

Reverts #4378, fixing ops-spotted bugs in https://discord.com/channels/554694662431178782/1103057209065218191/1103069753465114674 - mostly the same, with added testing steps for receive, fiat amount reactivity, and closing working and not regressing on regular sends.

Adds a dashboard-level QR code button, allowing to scan either an address, or an EVM/UTXO URL.

While at it, cleans up the send domain:

- removes `SendFormFields.Asset` in profit of `SendFormFields.AssetId`
- fixes occurences of `feeAssetId` currently being the `assetId` (i.e the asset being sent), which is obviously wrong and will produce errors
- removes useless `SendFormFields.CryptoSymbol` i.e the symbol of the asset being sent - we can get this from the `Asset` type

## Pull Request Type


## Issue (if applicable)

closes https://github.com/shapeshift/web/issues/3464
## Risk



Changes the API of:

- `estimateFees` consumed by bridging, send, Cosmos staking deposits
- `SendInput`, consumed by send, and THORChain Savers deposit/withdraws
- `<ReceiveRouter />`, i.e the receive modal routing

## Testing



- Bridging shows no regressions
- Cosmos staking deposits show no regressions
- Regular sends (i.e from account/asset page) show no regressions
- The QR code button works for
  - ETH addresses, as well as URLs with and without an amount (routing to asset selection, as we can't *safely* infer which asset is trying to be sent). Amounts should be populated in case the URL contains an `amount` parameter
  - UTXO addresses, as well as URLs with and without an amount - routing directly to the amount step, pre-populated in case an amount is present in the URL
- Ensure validation still works, both for QR-code-populated amount as well as user-input amounts. Try to input different amounts and ensure the validation is still reactive.
- Ensure fiat amount/crypto amount toggling still works, both on regular sends and QR code sends, with and without an amount set in the URL. Ensure keyboard input is still working after toggling back and forth and produces the right crypto/fiat amount.
- Ensure the QR code modal closes on Tx success/failure
- Ensure the receive modal still opens when clicking "Receive" from dashboard, and the modal is showing no regressions
  
### Engineering



- Implementation looks sane
- Refactor looks sane

### Operations



## Screenshots (if applicable)

<img width="1314" alt="image" src="https://user-images.githubusercontent.com/17035424/235537956-8fc79e91-8fb9-4564-ad04-a9e51c8848f9.png">
